feat: minimalistic web UI

timvahlbrock opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/steilerDev/icloud-photos-sync/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

fixes #638 

### ToDo's

- [x] Write some new tests, as the old ones aren't applicable anymore
- [x] Think about wether the app should exit when the server can't be started (currently it doesn't)
- [x] Documentation
- [x] Show state
- [x] Better control which views can be shown when
- [x] Allow pasting the mfa code
- [x] Handle Re-authentication when authentication is still valid
- [x] ~~wait for mfa result~~ (not possible, because mfa process fails on first failure)
  - [x] ~~before redirecting away from mfa submission~~
  - [x] ~~stay if the code is wrong and allow reentering it~~
- [x] allow mfa re-authentication if daemon is running with 'fail-on-mfa'
- [x] if ui should represent correct sync state with non-daemon apps, modify event handling accordingly

timvahlbrock commented:
I checked whether we might be able to fill in the code automatically, but it seems unlikely because we probably can't associate the `apple.com` domain with the web ui.

steilerDev commented:
> I checked whether we might be able to fill in the code automatically, but it seems unlikely because we probably can't associate the apple.com domain with the web ui.

I would be highly concerned if we could - this would enable phising the code :D

timvahlbrock commented:
Hey @steilerDev a little update on this: from my side the development of the web-ui itself is pretty much done. There is still some cleaning up to do (creating tests for the new functionality, documentation, going through the changes myself once again, so on), but if you already want to check out how the web ui turned out and want to give feedback on that, feel free to do so. I think I will be able to finish what I have on my ToDo's for this in the next week or during the easter holidays.

steilerDev commented:
AWESOME! Thank you so much for the work on this - I'll check it out this week and will comment (in case I have anything) 🤌

timvahlbrock commented:
Regarding the states shown: I wanted to keep the UI simple, so I kept it to the most important states for now.

timvahlbrock commented:
Hey @steilerDev I'm still having problems with some flaky tests, but other than that, this is done from my side.


Regarding the flaky tests: Somehow some tests seem to have their requests handled by the server instance of the previous instance. The postJson in the tests for reauthSuccess, and the two reauthError tests seem to fail the first postJson with ECONNRESET in the following tests, hence the retries and the waiting for a successful html request in the following suite. I spent some hours already debugging that, with little success. Maybe you spot something in the review. I've also got problems with the request mfa view tests since I added the 'paste' test in which the requests don't fail, but seem to only process the events from the past test. I will try to look further into both in the next to days, but the rest is already ready to be reviewed.

steilerDev commented:
Brilliant - thank you for the contribution and all the hard work! I hope I'll be able to get around checking it out this week! :)

steilerDev commented:
I think the `web-server.test.ts` should not perform actual web requests (is this the one with the flaky behaviour?) - I'll try and mock out the relevant functions and rework testing there.

For the web-ui I can see you went with JSDOM - maybe we should simply use playwright for actual UI tests... I'll play around with this (since I wanted to implement true WebUI tests for a while :D )

timvahlbrock commented:
> I think the `web-server.test.ts` should not perform actual web requests (is this the one with the flaky behaviour?) - I'll try and mock out the relevant functions and rework testing there.

I switched to starting an actual server to make the tests more integrative, but yeah, both of the flaky suites have troubles with this, so switching back should be an option. But I would still like to propose that we then use fake request/response objects that are handed in using a mocked http.server, instead of mocking and invoking (otherwise private) methods of the class under test, like before. 

> For the web-ui I can see you went with JSDOM - maybe we should simply use playwright for actual UI tests... I'll play around with this (since I wanted to implement true WebUI tests for a while :D )

Yeah, sounds reasonable, especially as you can see from some disabled tests jsdom doesn't seem to implement quite a lot browser interfaces. 

In a different context I also thought whether one could make the end to end testing easier. Setting up an own iCloud account for that is a little tedious and also requires manual interaction. Some things, like different shared library roles cannot even be tested with a single account. Maybe those could instead be run against a mocked http server and test the compliance with the iCloud API separately (I think the later is already the case?). This one probably also play nice with something like cucumber to describe the different scenarios and edge cases. This might play nice together with playwright, but as already said, for now playwright tests could surely run just as well without it.

timvahlbrock commented:
Found a library that at least mocks the Request and Response Objects and started implementing a mock for the http server based on that. Still need to clean this up though and need to apply it to the UI tests as well.

timvahlbrock commented:
So, I've now converted both the ui and the server tests to use a mocked instance of http.Server. I moved most of the http mocking implementations to the responsible class `MockedHttpServer` class and am pretty happy on how that turned out. I squashed the new commits into one, to avoid cluttering the default branch history with my rather small commits but still give an option to see what changed since the last time you took a look at it.

At the moment I am not too keen on switching the UI tests to playwright, but you mentioned that you wanted to give this a shot. Do you want to do this before merging this or in a separate PR?

steilerDev commented:
@timvahlbrock I wanted to look into this this week but ran into some local environment issues :D 
How stable are the tests at the moment? I don't *need* Playwright - but I'd like stable tests

steilerDev commented:
Perfect - then there is no need to overly complicate things! Could you please resolve the merge conflict from the updated upstream dependencies and then I'll try to merge before the next dependabot run this weekend 🤌🏻

timvahlbrock commented:
Done
